File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0210_Course_Schedule_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class Question_0210_Course_Schedule_II {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        if (prerequisites == null) {
            return new int[0];
        }
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<Integer>());
        }
        int[] inDegree = new int[numCourses];
        for (int[] pre : prerequisites) {
            int c = pre[0], p = pre[1];
            graph.get(p).add(c);
            inDegree[c]++;
        }
        Queue<Integer> que = new LinkedList<Integer>();
        for (int i = 0; i < inDegree.length; i++) {
            if (inDegree[i] == 0) {
                que.add(i);
            }
        }
        List<Integer> list = new ArrayList<>();
        while (!que.isEmpty()) {
            int i = que.poll();
            list.add(i);
            for (int j : graph.get(i)) {
                inDegree[j]--;
                if (inDegree[j] == 0) {
                    que.add(j);
                }
            }
        }
        for (int i : inDegree) {
            if (i != 0) {
                return new int[0];
            }
        }
        int[] res = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
            res[i] = list.get(i);
        }
        return res;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0210_Course_Schedule_II.cc
#include <vector>
#include <queue>

using namespace std;

class Solution {
 public:
  vector<int> findOrder(int numCourses, vector<pair<int, int>>& prerequisites) {
    vector<vector<int>> graph(numCourses, vector<int>(0));
    vector<int> inDegree(numCourses, 0);
    for (auto u : prerequisites) {
      graph[u.second].push_back(u.first);
      ++inDegree[u.first];
    }
    queue<int> que;
    for (int i = 0; i < numCourses; ++i) {
      if (inDegree[i] == 0) que.push(i);
    }
    vector<int> res;
    while (!que.empty()) {
      int u = que.front();
      que.pop();
      res.push_back(u);
      for (auto v : graph[u]) {
        --inDegree[v];
        if (inDegree[v] == 0) que.push(v);
      }
    }
    for (int i = 0; i < numCourses; ++i) {
      if (inDegree[i] != 0) return vector<int>();
    }
    return res;
  }
};
